py_ecc/bls/utils.py

```diff
@@ -66,31 +73,62 @@ def modular_squareroot_in_FQ2(value: FQ2) -> FQ2:
     return None
 
 
-def _get_x_coordinate(message_hash: Hash32, domain: Domain) -> FQ2:
-    # Initial candidate x coordinate
-    x_re = big_endian_to_int(hash_eth2(message_hash + domain + b'\x01'))
-    x_im = big_endian_to_int(hash_eth2(message_hash + domain + b'\x02'))
-    x_coordinate = FQ2([x_re, x_im])  # x_re + x_im * i
+def hash_to_G2(message_hash: Hash32) -> G2Uncompressed:
+    return hash_to_curve_G2(message_hash)
 
-    return x_coordinate
 
+def hash_to_curve_G2(message_hash: Hash32) -> G2Uncompressed:
+    u0 = hash_to_base_FQ2(message_hash, 0)
+    u1 = hash_to_base_FQ2(message_hash, 1)
+    q0 = map_to_curve_G2(u0)
+    q1 = map_to_curve_G2(u1)
+    r = q0 + q1
+    # TODO: Implement the following functions
+    # p = clear_cofactor_G2(r)
+    return r
 
-def hash_to_G2(message_hash: Hash32, domain: Domain) -> G2Uncompressed:
-    x_coordinate = _get_x_coordinate(message_hash, domain)
 
-    # Test candidate y coordinates until a one is found
-    while 1:
-        y_coordinate_squared = x_coordinate ** 3 + FQ2([4, 4])  # The curve is y^2 = x^3 + 4(i + 1)
-        y_coordinate = modular_squareroot_in_FQ2(y_coordinate_squared)
-        if y_coordinate is not None:  # Check if quadratic residue found
-            break
-        x_coordinate += FQ2([1, 0])  # Add 1 and try again
+def hash_to_base_FQ2(message_hash: Hash32, ctr: int) -> FQ2:
+    """
+    Hash To Base for FQ2
 
-    return multiply(
-        (x_coordinate, y_coordinate, FQ2([1, 0])),
-        G2_cofactor,
-    )
+    Converts a message to a point in the finite field as defined here:
+    https://tools.ietf.org/html/draft-irtf-cfrg-hash-to-curve-04#section-5
+    """
+    m_prime = hkdf_extract(DST, message_hash)
+    e = []
+
+    for i in range(1, 3):
+        # Concatenate ("H2C" || I2OSP(ctr, 1) || I2OSP(i, 1))
+        info = b'H2C' + bytes([ctr])[:1] + bytes([i])[:1]
```

Comment:
I think the slices here are un-necessary.  I believe the following statement is equivalent:

```
info = b'H2C' + bytes([ctr]) + bytes([i])
```

Reply:
I added these as insurance that `bytes([ctr])[:1] + bytes([i])[:1]` can oly ever be 1 byte.

You're right about `[i][:1]` that can be removed since i is either 1 or 2. However for `ctr` its type is `int`so to prevent accidental (or deliberate) misuse I thought it appropriate to ensure it is a slice of length 1. If you think this is overkill I'm happy to remove the slice.

Unless there is a way in the function to change `ctr` type to a single byte or u8?
`def hash_to_base_FQ2(message_hash: Hash32, ctr: int) -> FQ2:`

Reply:
```python
----> 1 bytes([256])

ValueError: bytes must be in range(0, 256)
```

I don't think it's possible for `bytes([ctr])` to be longer than a single byte.

Reply:
You're right I'll remove the slicing :)